#7190 Added sort_by config to ls configs (Fixes #7190)

RobbingDaHood opened the pull request:
# Description
- This PR closes #7190 

First of, first PR and I have some questions about missing pieces. I do not know if this is the best way to communicate it so please direct me. I added a questions section at the end. 

# User-Facing Changes
I am quite sure I solely use new code and beside adding the `sort_by`-configs then there should not be any other changes. 

# Tests + Formatting

```                                                                                            
~/Projects/nushell> cd tests/fixtures/completions/          
~/Projects/nushell/tests/fixtures/completions> ls      
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ another              │ dir  │ 4.0 KiB │ 2 days ago │
│ 1 │ custom_completion.nu │ file │   100 B │ 2 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 2 days ago │
│ 3 │ nushell              │ file │     0 B │ 2 days ago │
│ 4 │ test_a               │ dir  │ 4.0 KiB │ 2 days ago │
│ 5 │ test_b               │ dir  │ 4.0 KiB │ 2 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯

# Add `sort_by: ['name reverse natural', 'type']` to ls in `config.nu` and source. 
~/Projects/nushell/tests/fixtures/completions> ls                 
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ test_b               │ dir  │ 4.0 KiB │ 2 days ago │
│ 1 │ test_a               │ dir  │ 4.0 KiB │ 2 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 2 days ago │
│ 3 │ another              │ dir  │ 4.0 KiB │ 2 days ago │
│ 4 │ nushell              │ file │     0 B │ 2 days ago │
│ 5 │ custom_completion.nu │ file │   100 B │ 2 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯

# Remove the line again and source again (ERROR, it does not reset)
~/Projects/nushell/tests/fixtures/completions> ls                 
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ test_b               │ dir  │ 4.0 KiB │ 2 days ago │
│ 1 │ test_a               │ dir  │ 4.0 KiB │ 2 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 2 days ago │
│ 3 │ another              │ dir  │ 4.0 KiB │ 2 days ago │
│ 4 │ nushell              │ file │     0 B │ 2 days ago │
│ 5 │ custom_completion.nu │ file │   100 B │ 2 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯     

# Add `sort_by: []` to ls in `config.nu` and source. (Now it resets)
~/Projects/nushell/tests/fixtures/completions> ls      
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ another              │ dir  │ 4.0 KiB │ 2 days ago │
│ 1 │ custom_completion.nu │ file │   100 B │ 2 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 2 days ago │
│ 3 │ nushell              │ file │     0 B │ 2 days ago │
│ 4 │ test_a               │ dir  │ 4.0 KiB │ 2 days ago │
│ 5 │ test_b               │ dir  │ 4.0 KiB │ 2 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```  


# I did run the following 
- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library


# After Submitting
I need to update [the documentation](https://github.com/nushell/nushell.github.io) and I will prepare that while waiting for this MR to get some comments. 

# Questions 

1. Is there a better way of parsing the config than the new `setup_and_sort_by_config`-method? I kept it super naive just to be able to test the setup.  I were thinking that maybe there is already some parsing logic that parses strings into sort setup? Please direct me to some example code if that is the case. 
2. If that is not the case: Should I just keep it simple and have the fully written out flags? Because writing a parser that checks for `-in` as an example is fine but adds complexity. 
3. I were not able to find any existing automatic tests for `ls`, please direct me to them because I want to add some automatic tests for this feature. 
4. In general I did hack this together so I am very open for input about better architecture. I were keeping it simple so I could share quick and then get good feedback. 
5. Is it a general sourcing error that a config is not overwritten by default value if it is removed? As shown in the test. 
6. Is it okay that I removed the `Copy`-trait from `LsConfig`?

fdncred commented:
Seems interesting. There probably needs to be an entry in default_config.nu with documentation comments that explain what the configuration option is and how it works.

132ikl commented on crates/nu-command/src/filesystem/ls.rs:
```diff
@@ -219,6 +225,69 @@ impl Command for Ls {
+        let contains_naturel = sort_by_config.contains("natural"); // TODO expand
+        let contains_reverse = sort_by_config.contains("reverse"); // TODO expand
+
+        let mut compare_err: Option<ShellError> = None;
+        comparator_vals.sort_by(|a, b| {
+            compare_by(
+                a,
+                b,
+                &mut comparators,
+                call.head,
+                contains_insensitive,
+                contains_naturel,
+                &mut compare_err,
+            )
+        });
```
Could you use `sort_utils::sort_by` here instead? It does the same thing for you, less code duplication 😄

132ikl commented on crates/nu-command/src/filesystem/ls.rs:
```diff
@@ -219,6 +225,69 @@ impl Command for Ls {
     }
 }
 
+impl Ls {
```
These don't need to be in an `impl Ls` block, since they don't use `self` in the function signature.

RobbingDaHood commented:
> Seems interesting. There probably needs to be an entry in default_config.nu with documentation comments that explain what the configuration option is and how it works.

Done :)

132ikl commented on crates/nu-utils/src/sample_config/default_config.nu:
```diff
@@ -152,6 +152,7 @@ $env.config = {
     ls: {
         use_ls_colors: true # use the LS_COLORS environment variable to colorize output
         clickable_links: true # enable or disable clickable links. Your terminal has to support links.
+        sort_by: [] # Adds sorting on the output of ls, Example: sort_by: ['name reverse natural', 'type'] would be similar to `ls | | sort-by name -rn | sort-by type`
```
Could this be something more structured instead? Then you don't have to split by whitespace. Maybe something like this:

```nushell
sort_by: [
  {column: "name", reverse: true, natural: true},
  {column: "type"} # default reverse, natural, and insensitive to false 
]
```

132ikl commented on crates/nu-command/src/filesystem/ls.rs:
```diff
@@ -219,6 +225,69 @@ impl Command for Ls {
+        call: &Call,
+        result: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        let sort_by_configs = stack.get_config(engine_state).ls.sort_by.clone();
```
I think you should be able to avoid this clone, since you only pass it by reference later

132ikl commented:
I think this is a cool addition! Thanks for contributing 😄 

For (1) and (2), I think these could both be solved by using a record to specify the parameters (see my review comment).

For (3), see [`crates/nu-command/tests/commands/ls.rs`](https://github.com/nushell/nushell/tree/main/crates/nu-command/tests/commands/ls.rs)

For (5), I'm not sure. There's been some changes related to how config value defaults are handled, I thought a default value was supposed to be set if you construct a new `$env.config`, but not sure (cc @IanManske @NotTheDr01ds)

For (6), seems fine to me if it still works without it 😆

fdncred commented on crates/nu-protocol/src/config/ls.rs:
```diff
@@ -1,17 +1,19 @@
         Self {
             use_ls_colors: true,
             clickable_links: true,
+            sort_by: vec![],
```
https://github.com/nushell/nushell/pull/13875 enforces a sort order if landed. Lots of streaming concerns there but I have the same concerns here. I haven't looked closely enough to see if we're collecting more in this PR.

132ikl replied:
Hm, we probably shouldn't merge both #13875 and this. Double sorting (pre-sort from #13875, post-sort from this PR) isn't ideal. At the very least, if #13875 is merged, the default should probably be no post-sorting.

fdncred commented on crates/nu-command/src/filesystem/ls.rs:
```diff
@@ -219,6 +225,69 @@ impl Command for Ls {
+                Span::unknown(),
+            )],
+        })];
+        let contains_insensitive = sort_by_config.contains("ignore-case"); // TODO expand
```
what do the TODOs here mean?

RobbingDaHood replied:
It were just temporary to make it clear I wanted input. I did get the input and dig a new implementation.

fdncred commented on crates/nu-utils/src/sample_config/default_config.nu:
```diff
@@ -152,6 +152,7 @@ $env.config = {
     ls: {
         use_ls_colors: true # use the LS_COLORS environment variable to colorize output
         clickable_links: true # enable or disable clickable links. Your terminal has to support links.
+        sort_by: [] # Adds sorting on the output of ls, Example: sort_by: ['name reverse natural', 'type'] would be similar to `ls | | sort-by name -rn | sort-by type`
```
I'm also wondering if `sort-by name -rn | sort-by type` accurate. I thought sort-by could now sort multiple columns at the same time versus sort then sort again. There was a PR recently that overhauled the sorting to make it better.

132ikl replied:
You're right actually, I missed this. The implementation of this PR will do the equivalent of `sort-by -rn name type`, which will sort naturally for both name and type and then will reverse. You can't specify per comparator natural or reverse, you would have to run multiple `sort_by`s of one comparator each.

RobbingDaHood commented:
# Tests + Formatting

```
~/Projects/nushell> cd tests/fixtures/completions/      
```

```                                                                                                
~/Projects/nushell/tests/fixtures/completions> ls      
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ another              │ dir  │ 4.0 KiB │ 2 days ago │
│ 1 │ custom_completion.nu │ file │   100 B │ 2 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 2 days ago │
│ 3 │ nushell              │ file │     0 B │ 2 days ago │
│ 4 │ test_a               │ dir  │ 4.0 KiB │ 2 days ago │
│ 5 │ test_b               │ dir  │ 4.0 KiB │ 2 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```      

## Complex case 
``` 
        sort_by: [
            {
                column: "name",
                reverse: true,
                natural: true,
            },
            {
                column: "type",
            },
            {
                column: "size",
                reverse: true,
                natural: true,
            },
        ]     
```      
```      
~/Projects/nushell/tests/fixtures/completions> ls                                                                                                                         11/14/2024 07:49:47 PM
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ another              │ dir  │ 4.0 KiB │ 3 days ago │
│ 1 │ directory_completion │ dir  │ 4.0 KiB │ 3 days ago │
│ 2 │ test_a               │ dir  │ 4.0 KiB │ 3 days ago │
│ 3 │ test_b               │ dir  │ 4.0 KiB │ 3 days ago │
│ 4 │ custom_completion.nu │ file │   100 B │ 3 days ago │
│ 5 │ nushell              │ file │     0 B │ 3 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```      

## missing column 
``` 
        sort_by: [
            {
                reverse: true,
                natural: true,
            },
        ]     
```      

```      
~/Projects/nushell/tests/fixtures/completions> source ~/.config/nushell/config.nu                                                                                         11/14/2024 07:51:57 PM
Error: nu::shell::invalid_config

  × Encountered 1 error(s) when updating config

Error: nu::shell::missing_required_column

  × $env.config.ls.sort_by requires a 'column' column
     ╭─[~/.config/nushell/config.nu:156:13]
 155 │             sort_by: [
 156 │ ╭─▶             {
 157 │ │                   reverse: true,
 158 │ │                   natural: true,
 159 │ ├─▶             },
     · ╰──── has no 'column' column
 160 │                 {
     ╰────
```      

## invalid field 
``` 
        sort_by: [
            {
                column: "type",
                reverse: true,
                natural: true,
                invalidfield: true,
            },
        ]     
```      

```      
~/Projects/nushell/tests/fixtures/completions> source ~/.config/nushell/config.nu                                                                                         11/14/2024 07:53:13 PM
Error: nu::shell::invalid_config

  × Encountered 1 error(s) when updating config

Error: nu::shell::unknown_config_option

  × Unknown config option: $env.config.ls.sort_by.invalidfield
     ╭─[~/.config/nushell/config.nu:160:31]
 159 │                 natural: true,
 160 │                 invalidfield: true,
     ·                               ──┬─
     ·                                 ╰── remove this
 161 │             },
     ╰────
```      



# I did run the following 
- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library


# After Submitting
I need to update [the documentation](https://github.com/nushell/nushell.github.io) and I will prepare that while waiting for this MR to get some comments. 

# Questions 
All my questions in the description were answered, but I think I still need some help for automatic tests. 

The tests in `crates/nu-command/tests/commands/ls.rs` looks great, but can I change the configs before running the commands? I could not spot a case where that happened: Or is there any other good suggestion for test setup?

RobbingDaHood commented on crates/nu-command/tests/commands/ls.rs:
```diff
@@ -430,6 +430,28 @@ fn lists_files_including_starting_with_dot() {
 }
 
+#[test]
+fn lists_regular_files_using_question_mark_wildcard_asdsad() {
```
Ups, that need to be removed, it were just some experimental code. 

Will wait until I get some input on how to setup the automatic tests before changing this.

RobbingDaHood replied:
Fixed and pushed.

RobbingDaHood commented on crates/nu-protocol/src/config/ls.rs:
```diff
@@ -1,17 +1,82 @@
+                        match col.as_str() {
+                            "column" => ls_config_sort_config.column.update(val, path, errors),
+                            "reverse" => ls_config_sort_config.reverse.update(val, path, errors),
+                            "ignore_case" => {
```
I will remove the unnecessary `{` at next push.

RobbingDaHood replied:
This is because the line is so long, so cannot change that.

RobbingDaHood commented:
## invalid column (Needs improvement) 
``` 
        sort_by: [
            {
                column: "invalidcolumn",
                reverse: true,
                natural: true,
            },
        ]     
```      

```      
~/Projects/nushell> source ~/.config/nushell/config.nu                                                                                                                    11/14/2024 08:30:50 PM
~/Projects/nushell> ls                                                                                                                                                    11/14/2024 08:30:54 PM
Error: nu::shell::column_not_found

  × Cannot find column 'invalidcolumn (from the ls configuration)'
   ╭─[source:1:1]
 1 │ nu
   · ▲
   · ╰── cannot find column 'invalidcolumn (from the ls configuration)'
   ╰────
   ╭─[entry #7:1:1]
 1 │ ls
   · ─┬
   ·  ╰── value originates here
   ╰────
```      

Is that too hacky? :D

fdncred commented:
Now that the other `ls` with default sort PR has landed, we really don't want to be double sorting. You might want to rebase and see if there's an easy way to make this work again.

RobbingDaHood commented:
> Now that the other `ls` with default sort PR has landed, we really don't want to be double sorting. You might want to rebase and see if there's an easy way to make this work again.

@fdncred Would it be to aggressive to just revert those changes? https://github.com/nushell/nushell/pull/13875/files 

Then I can add the default sorting on name mentioned further up in this MR: https://github.com/nushell/nushell/pull/14331#discussion_r1840890846 

There there is still a default order, it is configurable and it is also enabled in multi-threading.

RobbingDaHood commented:
> > Now that the other `ls` with default sort PR has landed, we really don't want to be double sorting. You might want to rebase and see if there's an easy way to make this work again.
> 
> @fdncred Would it be to aggressive to just revert those changes? https://github.com/nushell/nushell/pull/13875/files
> 
> Then I can add the default sorting on name mentioned further up in this MR: [#14331 (comment)](https://github.com/nushell/nushell/pull/14331#discussion_r1840890846)
> 
> There there is still a default order, it is configurable and it is also enabled in multi-threading.

I just pushed those changes: 

```
    ls: {
        use_ls_colors: true # use the LS_COLORS environment variable to colorize output
        clickable_links: true # enable or disable clickable links. Your terminal has to support links.
    }
```  

```
~/Projects/nushell/tests/fixtures/completions> ls         
╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ another              │ dir  │ 4.0 KiB │ 5 days ago │
│ 1 │ custom_completion.nu │ file │   100 B │ 5 days ago │
│ 2 │ directory_completion │ dir  │ 4.0 KiB │ 5 days ago │
│ 3 │ nushell              │ file │     0 B │ 5 days ago │
│ 4 │ test_a               │ dir  │ 4.0 KiB │ 5 days ago │
│ 5 │ test_b               │ dir  │ 4.0 KiB │ 5 days ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```

132ikl commented:
At the last meeting someone brought up that this PR would make command output (`ls`'s, specifically) dependent on config, and that's something I hadn't considered before. I do think that this PR is better for UX than just always sorting in the same way, but we don't have very well defined semantics for what config "should" or "should not" be able to do. This "invalid column" example you posted is a good example of how this can cause issues:

> ~/Projects/nushell> ls                                                                                                                                                    11/14/2024 08:30:54 PM
> Error: nu::shell::column_not_found
> 
>   × Cannot find column 'invalidcolumn (from the ls configuration)'
>    ╭─[source:1:1]
>  1 │ nu
>    · ▲
>    · ╰── cannot find column 'invalidcolumn (from the ls configuration)'
>    ╰────
>    ╭─[entry #7:1:1]
>  1 │ ls
>    · ─┬
>    ·  ╰── value originates here
>    ╰────
> ```

I will note that `$env.config.rm.always_trash` is an example of another config option that changes actual command behavior (rather than just output formatting). I think I am still in favor of this PR, but I would prefer that failure to sort doesn't cause `ls` to break. Silently failing and not sorting the output isn't great either though, so I'm not sure what the best way to handle this would be.

fdncred commented:
I agree. I don't like the failure. Maybe a warning, "Invalid blah blah blah, defaulting to sort by name".

RobbingDaHood commented:
> I agree. I don't like the failure. Maybe a warning, "Invalid blah blah blah, defaulting to sort by name".

I tried adding a `            warn! {"{} (from the ls configuration)", col_name};` but that did not do a difference, but maybe I can enable seeing warnings. I this something you were thinking? 

I do think it is easier and maybe more safe to just ignore the sort in this case: There could be multiple sorts in a row and defaulting to anything could be even more distracting? 

Also just as a reminder, then I am still looking for input on how to automatic test this change in a good way: 

> All my questions in the description were answered, but I think I still need some help for automatic tests.
> 
> The tests in crates/nu-command/tests/commands/ls.rs looks great, but can I change the configs before running the commands? I could not spot a case where that happened: Or is there any other good suggestion for test setup?

fdncred commented:
> I tried adding a warn!  ....

I wasn't talking about a log warning. I meant a nushell warning. Nushell has ShellErrors and there are a few warnings like when commands are deprecated. They still work but warn you that they'll soon be removed.

> I do think it is easier and maybe more safe to just ignore the sort in this case ...

I don't like the idea of failing silently.

132ikl commented:
> can I change the configs before running the commands?

It looks like you might be able to use the `Playground::with_config` method? That requires an _actual_ config file, but you might be able to use `nu_test_support::fs::Stub::FileWithContent` for that? I don't think there are any tests which use config right now, the only test which uses the `Playground::with_config` method is `crates/nu-command/tests/commands/source_env.rs` and, uh, I'm not really sure what's that's doing with its config (`.toml`??)

RobbingDaHood commented:
> > I tried adding a warn!  ....
> 
> I wasn't talking about a log warning. I meant a nushell warning. Nushell has ShellErrors and there are a few warnings like when commands are deprecated. They still work but warn you that they'll soon be removed.
> 
> > I do think it is easier and maybe more safe to just ignore the sort in this case ...
> 
> I don't like the idea of failing silently.

1. `ShellError::Deprecated` is an error like all the others. Just tried using that error instead. 
2. `ParseWarning::DeprecatedWarning` does not seem to be used anywhere. This does not seem to be a parse warning either. 
3. The `Command::run` trait depends on returning a `Result<PipelineData, ShellError>`. 
4. Should I make a new `ShellWarning` enum? And implement the code to handle that? Looks like that would total rewrite the `Command::run` because it expects `ShellError`? Quickest rewrite is likely defining a `NonOkEnum` that can be either `ShellError` or `ShellWarning`? But that seems like major too. 
5. Maybe keeping the hard error in there for now and then create an issue on supporting `ShellWarning` can be done later? Then we can discuss a good setup for it? Mainly because the hard error will be triggered the first time ls is executed. 
6. Then we just need to agree on a text for the current implemented hard error.

RobbingDaHood commented:
# I added some tests! :tada: 

Hope they follow the standard. 

# I improved the "bad column name"-config-error
```
        sort_by: [
            {
                column: "invalid",
                reverse: true,
                natural: true,
            },
        ]   
```  

``` 
~/Projects/nushell> ls                                                                                                                                                    11/19/2024 06:45:19 PM
Error: nu::shell::column_not_found

  × Cannot find column 'invalid (from the ls configuration, possible columns are name, type, size, modified.)'
   ╭─[source:1:1]
 1 │ nu
   · ▲
   · ╰── cannot find column 'invalid (from the ls configuration, possible columns are name, type, size, modified.)'
   ╰────
   ╭─[entry #4:1:1]
 1 │ ls
   · ─┬
   ·  ╰── value originates here
   ╰────
``` 

# It also works if no files are passed from the ls

```
~/Projects/nushell> ls testing                                                                                                                                            11/19/2024 06:51:01 PM
╭────────────╮
│ empty list │
╰────────────╯
```  

# The only open talk that is left 

Seems to be if it should be a warning if the config is wrong. 

Even if adding this as a warning seems better then it is quite a lot of work to implement, as far as I can see. I wrote a bit about it here https://github.com/nushell/nushell/pull/14331#issuecomment-2483541355

fdncred commented:
I believe the warning I was talking about is report_shell_error like Wind is using in this PR https://github.com/nushell/nushell/pull/14385

RobbingDaHood commented:
# Now a misconfigured column name results in a warning 

```
~/Projects/nushell/tests/fixtures/completions> ls                                                                                                                         11/20/2024 07:39:19 AM
Warning: nu::shell::column_not_found

  × Cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, size, modified. Will ignore that misconfigured sort.)'
   ╭─[source:1:1]
 1 │ nu
   · ▲
   · ╰── cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, size, modified. Will ignore that misconfigured sort.)'
   ╰────
   ╭─[entry #6:1:1]
 1 │ ls
   · ─┬
   ·  ╰── value originates here
   ╰────

╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ test_b               │ dir  │ 4.0 KiB │ a week ago │
│ 1 │ test_a               │ dir  │ 4.0 KiB │ a week ago │
│ 2 │ nushell              │ file │     0 B │ a week ago │
│ 3 │ directory_completion │ dir  │ 4.0 KiB │ a week ago │
│ 4 │ custom_completion.nu │ file │   100 B │ a week ago │
│ 5 │ another              │ dir  │ 4.0 KiB │ a week ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```

RobbingDaHood commented:
# Status 

As far as I can see then I have handled all inputs right? 

Thx for all the good comments and please do not hesitate with more comments. :heart:

fdncred commented:
> Warning: nu::shell::column_not_found
```
  × Cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, size, modified. Will ignore that misconfigured sort.)'
   ╭─[source:1:1]
 1 │ nu
   · ▲
   · ╰── cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, size, modified. Will ignore that misconfigured sort.)'
   ╰────
   ╭─[entry #6:1:1]
 1 │ ls
   · ─┬
   ·  ╰── value originates here
   ╰────
```
Is it possible to make this not repeat the same warning string? I also wonder why the span is pointing at nu and why we cannot know what the "unknown column" name is.

132ikl commented:
I'll take a closer look later today, but I did notice you're creating the valid columns list at call time. I'm a bit concerned that if you did `ls -l` and had an invalid column in your config, the error would suggest using all of the columns from `ls -l`, when probably only name, type, size, and modified should be suggested since those are the only ones that are always there (is there a canonical list of columns shown by `ls` without `-l`?)

Also, I would like to hear from some of the core team members about what they think about the config changes, I wouldn't want you to put this much work into this PR and then get rejected at the last minute. Not sure if I can make it yet, but @fdncred maybe a topic for today's meeting?

RobbingDaHood commented:
> Is it possible to make this not repeat the same warning string? 

Yes :)  

``` 
~/Projects/nushell/tests/fixtures/completions> ls                                                                                                                         11/20/2024 05:16:26 PM
Warning: nu::shell::column_not_found

  × Cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, size, modified. Will ignore that misconfigured sort.)'
   ╭─[entry #9:1:1]
 1 │ ls
   · ─┬
   ·  ╰── value originates here
   ╰────

╭───┬──────────────────────┬──────┬─────────┬────────────╮
│ # │         name         │ type │  size   │  modified  │
├───┼──────────────────────┼──────┼─────────┼────────────┤
│ 0 │ test_b               │ dir  │ 4.0 KiB │ a week ago │
│ 1 │ test_a               │ dir  │ 4.0 KiB │ a week ago │
│ 2 │ nushell              │ file │     0 B │ a week ago │
│ 3 │ directory_completion │ dir  │ 4.0 KiB │ a week ago │
│ 4 │ custom_completion.nu │ file │   100 B │ a week ago │
│ 5 │ another              │ dir  │ 4.0 KiB │ a week ago │
╰───┴──────────────────────┴──────┴─────────┴────────────╯
```  

Will push fix later.

RobbingDaHood commented:
> the error would suggest using all of the columns from ls -l 

True, does not look that good: 

```
~/Projects/nushell/tests/fixtures/completions> ls -l                                                                                                                      11/20/2024 05:16:28 PM
Warning: nu::shell::column_not_found

  × Cannot find column 'unknown_column (from the ls configuration, possible columns are name, type, target, readonly, mode, num_links, inode, user, group, size, created, accessed, modified.
  │ Will ignore that misconfigured sort.)'
   ╭─[entry #10:1:1]
 1 │ ls -l
   · ─┬
   ·  ╰── value originates here
   ╰────

╭────┬──────────────────────┬──────┬────────┬──────────┬───────────┬───────────┬──────────┬─────────────────────────┬─────────────────────────┬─────────┬────────────┬────────────┬────────────╮
│  # │         name         │ type │ target │ readonly │   mode    │ num_links │  inode   │          user           │          group          │  size   │  created   │  accessed  │  modified  │
├────┼──────────────────────┼──────┼────────┼──────────┼───────────┼───────────┼──────────┼─────────────────────────┼─────────────────────────┼─────────┼────────────┼────────────┼────────────┤
│  0 │ test_b               │ dir  │        │ false    │ rwxrwxr-x │         2 │ 17861606 │ anon │ anon │ 4.0 KiB │ a week ago │ a day ago  │ a week ago │
│  1 │ test_a               │ dir  │        │ false    │ rwxrwxr-x │         2 │ 17861604 │ anon │ anon │ 4.0 KiB │ a week ago │ a day ago  │ a week ago │
│  2 │ nushell              │ file │        │ false    │ rw-rw-r-- │         1 │ 17861603 │ anon │ anon │     0 B │ a week ago │ 2 days ago │ a week ago │
│  3 │ directory_completion │ dir  │        │ false    │ rwxrwxr-x │         3 │ 17861599 │ anon │ anon │ 4.0 KiB │ a week ago │ a day ago  │ a week ago │
│  4 │ custom_completion.nu │ file │        │ false    │ rw-rw-r-- │         1 │ 17861598 │ anon │ anon │   100 B │ a week ago │ 2 days ago │ a week ago │
│  5 │ another              │ dir  │        │ false    │ rwxrwxr-x │         2 │ 17861594 │ anon │ anon │ 4.0 KiB │ a week ago │ a day ago  │ a week ago │
╰────┴──────────────────────┴──────┴────────┴──────────┴───────────┴───────────┴──────────┴─────────────────────────┴─────────────────────────┴─────────┴────────────┴────────────┴────────────╯
```

fdncred commented:
After our team meeting today, it sounds like we're not going to be able to move forward with this PR. The position is that we don't want nushell to depend on the config to use the `ls` command. There were other concerns as well. I can't remember them all and don't see them on the meeting notes. Maybe others will chime in who were on the call.

Sorry for taking so long to come to this decision. I was really hoping it would move forward. We really appreciate you trying your very best to support nushell and make it better.

Ian posted this as a potential work-around. I haven't tested it yet.
```nushell
def lss [
    --all (-a),         # Show hidden files
    --long (-l),        # Get all available columns for each entry (slower; columns are platform-dependent)
    --short-names (-s), # Only print the file names, and not the path
    --full-paths (-f),  # display paths as absolute paths
    --du (-d),          # Display the apparent directory size ("disk usage") in place of the directory metadata size
    --directory (-D),   # List the specified directory itself instead of its contents
    --mime-type (-m),   # Show mime-type in type column instead of 'file' (based on filenames only; files' contents are not examined)
    --threads (-t),     # Use multiple threads to list contents. Output will be non-deterministic.
    ...pattern: glob,   # The glob pattern to use.
]: [ nothing -> table ] {
    let pattern = if ($pattern | is-empty) { [ '.' ] } else { $pattern }
    (ls
        --all=$all
        --long=$long
        --short-names=$short_names
        --full-paths=$full_paths
        --du=$du
        --directory=$directory
        --mime-type=$mime_type
        --threads=$threads
        ...$pattern
    ) | sort-by -i type name
}
```

RobbingDaHood commented:
> After our team meeting today, it sounds like we're not going to be able to move forward with this PR. The position is that we don't want nushell to depend on the config to use the ls command.  

Fair enough, thx for the answer. 

My main motivation is to write some rust code and learn: I achieved both in this PR even though it did not get merged :) 

From the first comments there were also warnings that this could happen, so it did not come out of the blue. I expect to jump on another issue soon :heart:
